### ISSUE TITLE
How to calculate Mc without BatchNorm Layer

### ISSUE THREAD

**Issue by hwc1824** [clovaai/overhaul-distillation#7]
Hi,
I notice that in you paper, you mention that

> margin can be obtained directly during training

Does this mean that I just inference every training image with teacher model and average all negative value in each channel?

Also, I'm not quite sure how does the channel margin work when it comes to CNN Kernel.
For example, here is a set of CNN kernel and its shape is [56, 56, 64] for Height, Width and Channel. Does this mean that I should have 64 channel-wise margin?

**Comment by bhheo** [comment]
Hi,

Yes. What you said is exactly right.
That mention mean average all negative value in each channel for every training image.
And in your example, the margin is 64-dim channel-wise vector.

Here is the example code for directly calculating margin.
It might help you to implement algorithm.
``` Python
with torch.no_grad():
   for ii in range(len(t_feat)):
      m = (t_feat[ii] * (t_feat[ii] < 0).float()).sum_exclude_channel_dim
      m = m / ((t_feat[ii] < 0).float().sum_exclude_channel_dim + 1e-10)
      if (iteration) == 0:
          margin.append(m)
      else:
          margin[ii] = margin[ii] * iteration / (iteration+1) + m / (iteration+1)
```

**Comment by hwc1824** [comment]
Thanks for your fast response and sample code.
I have some new questions about the code.

1. does `t_feat` mean list of feature  1) along with different place where the distilling process happened or 2) along with different iteration?. I assumed it's former though.
2. `margin[ii] = margin[ii] * iteration / (iteration+1) + m / (iteration+1)`, Is it some kind of moving average? I wonder if this is necessary if I can actually calculate the true mean by inference all image after teacher model has been trained already.

Also, I come up with another question is that my teacher model is ResNet-100 and the student is ResNet-34 and the feature map I choose to distill for the both network is in completely same shape (Height, Width and Channel). I wonder in such case, if the regressor (1x1 conv + batchnorm) for student transform is still necessary or not. 

Again, thank for your kind response

**Comment by bhheo** [comment]
1.
It is 1). features from different place at same iteration.

2.
It is just used to calculate margin with training process. It approximates true mean.
So, if you have the true mean, the true mean is better.

- Regressor for same shape.
As I remember, you can skip the regressor in same shape networks. 
Performance was slightly different.
So, it is not necessary for same shape.

**Comment by hwc1824** [comment]
Thanks for your kind response =D
